File: dark_multi/branch.py
"""Branch management for dark-multi."""

from datetime import datetime
from pathlib import Path

from .config import DARK_ROOT, OVERRIDES_DIR, Colors, run


class Branch:
    """Represents a branch clone."""

    def __init__(self, name: str):
        self.name = name
        self.path = DARK_ROOT / name
        self.override_dir = OVERRIDES_DIR / name
        self.metadata_file = self.override_dir / "metadata"

    @property
    def exists(self) -> bool:
        return self.path.is_dir() and (self.path / ".git").exists()

    @property
    def is_managed(self) -> bool:
        return self.override_dir.is_dir() and self.metadata_file.is_file()

    @property
    def metadata(self) -> dict:
        data = {}
        if self.metadata_file.is_file():
            for line in self.metadata_file.read_text().strip().split("\n"):
                if "=" in line:
                    k, v = line.split("=", 1)
                    data[k] = v
        return data

    @property
    def instance_id(self) -> int:
        return int(self.metadata.get("ID", 0))

    @property
    def container_name(self) -> str:
        return f"dark-{self.name}"

    @property
    def container_id(self) -> str | None:
        # Try by name first (new containers)
        result = run(["docker", "ps", "-q", "--filter", f"name=^{self.container_name}$"])
        cid = result.stdout.strip()
        if cid:
            return cid
        # Fall back to label (old containers)
        result = run(["docker", "ps", "-q", "--filter", f"label=dark-dev-container={self.name}"])
        cid = result.stdout.strip()
        return cid if cid else None

    @property
    def is_running(self) -> bool:
        return self.container_id is not None

    @property
    def has_changes(self) -> bool:
        if not self.exists:
            return False
        result = run(["git", "status", "--porcelain"], cwd=self.path)
        return bool(result.stdout.strip())

    @property
    def port_base(self) -> int:
        return 10011 + self.instance_id * 100

    @property
    def bwd_port_base(self) -> int:
        return 11001 + self.instance_id * 100

    def write_metadata(self, instance_id: int) -> None:
        self.override_dir.mkdir(parents=True, exist_ok=True)
        self.metadata_file.write_text(
            f"ID={instance_id}\n"
            f"NAME={self.name}\n"
            f"CREATED={datetime.now().isoformat()}\n"
        )

    def status_line(self) -> str:
        if self.is_running:
            status = f"{Colors.GREEN}running{Colors.NC}"
        else:
            status = f"{Colors.RED}stopped{Colors.NC}"
        changes = f" {Colors.YELLOW}[modified]{Colors.NC}" if self.has_changes else ""
        ports = f"ports {self.port_base}+/{self.bwd_port_base}+"
        return f"{Colors.BOLD}{self.name:20}{Colors.NC} {status:20} {ports}{changes}"


def find_next_instance_id() -> int:
    """Find the next available instance ID."""
    max_id = 0
    if OVERRIDES_DIR.is_dir():
        for path in OVERRIDES_DIR.iterdir():
            if path.is_dir():
                meta = path / "metadata"
                if meta.is_file():
                    for line in meta.read_text().split("\n"):
                        if line.startswith("ID="):
                            try:
                                max_id = max(max_id, int(line.split("=")[1]))
                            except:
                                pass
    return max_id + 1


def find_source_repo() -> Path | None:
    """Find a repo to clone from."""
    from .config import DARK_SOURCE

    # Check DARK_SOURCE
    if DARK_SOURCE != DARK_ROOT and DARK_SOURCE.is_dir() and (DARK_SOURCE / ".git").exists():
        return DARK_SOURCE

    # Check for 'main' branch
    main = DARK_ROOT / "main"
    if main.is_dir() and (main / ".git").exists():
        return main

    # Check any existing managed branch (via overrides dir)
    for branch in get_managed_branches():
        if branch.exists:
            return branch.path

    return None


def get_managed_branches() -> list[Branch]:
    """Get all managed branches by scanning overrides directory."""
    branches = []
    if OVERRIDES_DIR.is_dir():
        for path in sorted(OVERRIDES_DIR.iterdir()):
            if path.is_dir() and (path / "metadata").is_file():
                b = Branch(path.name)
                branches.append(b)
    return branches

File: branch/branch.go
// Package branch provides branch management for dark-multi.
package branch

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/darklang/dark-multi/config"
)

// Branch represents a branch clone.
type Branch struct {
	Name         string
	Path         string
	OverrideDir  string
	MetadataFile string
}

// New creates a new Branch instance.
func New(name string) *Branch {
	path := filepath.Join(config.DarkRoot, name)
	overrideDir := filepath.Join(config.OverridesDir, name)
	return &Branch{
		Name:         name,
		Path:         path,
		OverrideDir:  overrideDir,
		MetadataFile: filepath.Join(overrideDir, "metadata"),
	}
}

// Exists returns true if the branch directory exists with a .git folder.
func (b *Branch) Exists() bool {
	gitPath := filepath.Join(b.Path, ".git")
	info, err := os.Stat(gitPath)
	return err == nil && info.IsDir()
}

// IsManaged returns true if this is a dark-multi managed branch.
func (b *Branch) IsManaged() bool {
	// Check override dir exists and has metadata file
	info, err := os.Stat(b.OverrideDir)
	if err != nil || !info.IsDir() {
		return false
	}
	info, err = os.Stat(b.MetadataFile)
	return err == nil && !info.IsDir()
}

// Metadata returns the branch metadata as a map.
func (b *Branch) Metadata() map[string]string {
	data := make(map[string]string)
	content, err := os.ReadFile(b.MetadataFile)
	if err != nil {
		return data
	}
	for _, line := range strings.Split(string(content), "\n") {
		if idx := strings.Index(line, "="); idx > 0 {
			data[line[:idx]] = line[idx+1:]
		}
	}
	return data
}

// GetName returns the branch name (implements container.BranchInfo).
func (b *Branch) GetName() string {
	return b.Name
}

// GetPath returns the branch path (implements container.BranchInfo).
func (b *Branch) GetPath() string {
	return b.Path
}

// InstanceID returns the branch instance ID.
func (b *Branch) InstanceID() int {
	if id, ok := b.Metadata()["ID"]; ok {
		if i, err := strconv.Atoi(id); err == nil {
			return i
		}
	}
	return 0
}

// ContainerName returns the Docker container name.
func (b *Branch) ContainerName() string {
	return fmt.Sprintf("dark-%s", b.Name)
}

// ContainerID returns the running container ID, if any.
func (b *Branch) ContainerID() (string, error) {
	// Try by name first (new containers)
	cmd := exec.Command("docker", "ps", "-q", "--filter", fmt.Sprintf("name=^%s$", b.ContainerName()))
	out, err := cmd.Output()
	if err == nil {
		if id := strings.TrimSpace(string(out)); id != "" {
			return id, nil
		}
	}

	// Fall back to label (old containers)
	cmd = exec.Command("docker", "ps", "-q", "--filter", fmt.Sprintf("label=dark-dev-container=%s", b.Name))
	out, err = cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// IsRunning returns true if the container is running.
func (b *Branch) IsRunning() bool {
	id, err := b.ContainerID()
	return err == nil && id != ""
}

// HasChanges returns true if there are uncommitted changes.
func (b *Branch) HasChanges() bool {
	if !b.Exists() {
		return false
	}
	cmd := exec.Command("git", "-C", b.Path, "status", "--porcelain")
	out, err := cmd.Output()
	return err == nil && len(strings.TrimSpace(string(out))) > 0
}

// GitStatus returns modified and untracked file counts.
func (b *Branch) GitStatus() (modified int, untracked int) {
	if !b.Exists() {
		return 0, 0
	}
	cmd := exec.Command("git", "-C", b.Path, "status", "--porcelain")
	out, err := cmd.Output()
	if err != nil {
		return 0, 0
	}
	for _, line := range strings.Split(strings.TrimSpace(string(out)), "\n") {
		if line == "" {
			continue
		}
		if strings.HasPrefix(line, "??") {
			untracked++
		} else {
			modified++
		}
	}
	return modified, untracked
}

// GitStats returns commits ahead of main and total lines added/removed (committed + uncommitted).
func (b *Branch) GitStats() (commits int, added int, removed int) {
	if !b.Exists() || b.Name == "main" {
		return 0, 0, 0
	}

	// Try different refs to compare against
	refs := []string{"origin/main", "main"}
	var baseRef string
	for _, ref := range refs {
		cmd := exec.Command("git", "-C", b.Path, "rev-parse", "--verify", ref)
		if cmd.Run() == nil {
			baseRef = ref
			break
		}
	}

	if baseRef != "" {
		// Count commits ahead of base
		cmd := exec.Command("git", "-C", b.Path, "rev-list", "--count", baseRef+"..HEAD")
		out, err := cmd.Output()
		if err == nil {
			fmt.Sscanf(strings.TrimSpace(string(out)), "%d", &commits)
		}

		// Get total diff stats vs base (includes uncommitted)
		cmd = exec.Command("git", "-C", b.Path, "diff", "--numstat", baseRef)
		out, err = cmd.Output()
		if err == nil {
			for _, line := range strings.Split(string(out), "\n") {
				fields := strings.Fields(line)
				if len(fields) >= 2 {
					var a, r int
					fmt.Sscanf(fields[0], "%d", &a)
					fmt.Sscanf(fields[1], "%d", &r)
					added += a
					removed += r
				}
			}
		}
	} else {
		// No base ref found - just show uncommitted changes
		cmd := exec.Command("git", "-C", b.Path, "diff", "--numstat", "HEAD")
		out, err := cmd.Output()
		if err == nil {
			for _, line := range strings.Split(string(out), "\n") {
				fields := strings.Fields(line)
				if len(fields) >= 2 {
					var a, r int
					fmt.Sscanf(fields[0], "%d", &a)
					fmt.Sscanf(fields[1], "%d", &r)
					added += a
					removed += r
				}
			}
		}
	}

	return commits, added, removed
}

// PortBase returns the test port base for this branch.
func (b *Branch) PortBase() int {
	return 10011 + b.InstanceID()*100
}

// BwdPortBase returns the BwdServer port base for this branch.
func (b *Branch) BwdPortBase() int {
	return 11001 + b.InstanceID()*100
}

// WriteMetadata writes the branch metadata file.
func (b *Branch) WriteMetadata(instanceID int) error {
	if err := os.MkdirAll(b.OverrideDir, 0755); err != nil {
		return err
	}
	content := fmt.Sprintf("ID=%d\nNAME=%s\nCREATED=%s\n",
		instanceID, b.Name, time.Now().Format(time.RFC3339))
	return os.WriteFile(b.MetadataFile, []byte(content), 0644)
}

// StatusLine returns a formatted status line for display.
func (b *Branch) StatusLine() string {
	status := "stopped"
	if b.IsRunning() {
		status = "running"
	}
	changes := ""
	if b.HasChanges() {
		changes = " [modified]"
	}
	return fmt.Sprintf("%-20s %-10s ports %d+/%d+%s",
		b.Name, status, b.PortBase(), b.BwdPortBase(), changes)
}
